[WebAssembly] Recognise EXTEND_HIGH

sparker-arm opened the pull request:
When lowering EXTEND_VECTOR_INREG, check whether the operand is a shuffle that is moving the top half of a vector into the lower half. If so, we can EXTEND_HIGH the input to the shuffle instead.

tlively commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -2143,6 +2143,32 @@ WebAssemblyTargetLowering::LowerSIGN_EXTEND_INREG(SDValue Op,
+
+  // Look for a shuffle which moves from the high half to the low half.
+  size_t FirstIdx = Mask.size() / 2;
+  for (size_t i = 0; i < Mask.size() / 2; ++i) {
```
Do we have to check that the top half of the mask has expected values as well?

sparker-arm replied:
We don't, but as the user is an `extend_low`, the high half of the result doesn't matter. But I will add a check that the extend is the only user.

sparker-arm replied:
I implemented and then removed the user check as I couldn't see the benefit.

tlively commented on llvm/test/CodeGen/WebAssembly/extend-shuffles.ll:
```diff
@@ -0,0 +1,442 @@
+; SIMD128-NEXT:    return $pop8
+  %narrow.load = load <8 x i8>, ptr %in, align 1
+  %sext = sext <8 x i8> %narrow.load to <8 x i32>
+  %res = tail call i32 @llvm.vector.reduce.add.v8i32(<8 x i32> %sext)
```
Would it be possible to write tests that check the pattern more directly as well?

sparker-arm replied:
Fair enough, I'll give it a go.

sparker-arm replied:
The tests now just use a shuffle and an extend.

sparker-arm committed:
[WebAssembly] Recognise EXTEND_HIGH

When lowering EXTEND_VECTOR_INREG, check whether the operand is a
shuffle that is moving the top half of a vector into the lower half.
If so, we can EXTEND_HIGH the input to the shuffle instead.

ppenzin commented:
@jing-bao FYI

tlively reviewed (approved):
Looks great, thanks!

ppenzin commented on llvm/test/CodeGen/WebAssembly/extend-shuffles.ll:
```diff
@@ -0,0 +1,227 @@
+; SIMD128-LABEL: sext_high_v2i16:
+; SIMD128:         .functype sext_high_v2i16 (v128) -> (v128)
+; SIMD128-NEXT:  # %bb.0:
+; SIMD128-NEXT:    i8x16.shuffle $push0=, $0, $0, 4, 5, 6, 7, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1
+; SIMD128-NEXT:    i32x4.extend_low_i16x8_s $push1=, $pop0
+; SIMD128-NEXT:    return $pop1
+ %shuffle = shufflevector <4 x i16> %in, <4 x i16> poison, <2 x i32> <i32 2, i32 3>
+ %res = sext <2 x i16> %shuffle to <2 x i32>
```
Why won't extend_high get picked here? (and in couple more spots in this test)

sparker-arm replied:
Because the pattern matching requires shuffling the MSB eight bytes into the LSB eight bytes.
